### PR TITLE
Add unit tests

### LINKED ISSUE #7: Add unit tests for SqlRunner and KubernetesSecretReplacer
Also extracting the secret resolver as an interface and inject it into KubernetesSecretReplacer class would make testing a lot easier.

### PR DESCRIPTION
**Pull request by tinaselenge**
Closes #7

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -23,6 +23,10 @@
     <fabric8.kubernetes-client.version>6.13.0</fabric8.kubernetes-client.version>
     <flink.avro.confluent.registry.version>1.19.1</flink.avro.confluent.registry.version>
     <flink.kafka.connector.version>3.2.0-1.19</flink.kafka.connector.version>
+
+    <!-- Test only dependencies -->
+    <mockito.version>4.11.0</mockito.version>
+    <jupiter.version>5.8.2</jupiter.version>
```
Comment (Frawless) [inline pom.xml:29]:
```suggestion
    <jupiter.version>5.10.3</jupiter.version>
```

File: flink-sql-runner/src/test/java/com/github/streamshub/flink/KubernetesSecretReplacerImplTest.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    @BeforeEach
+    public void setUp() {
+        mockClient = mock(KubernetesClient.class);
+        MixedOperation mockSecrets = mock(MixedOperation.class);
+        Resource mockResource = mock(Resource.class);
+
+        // Set up the mock behavior
+        when(mockClient.secrets()).thenReturn(mockSecrets);
+        when(mockSecrets.inNamespace("default")).thenReturn(mockSecrets);
+        when(mockSecrets.withName("my-secret")).thenReturn(mockResource);
```
Comment (Frawless) [inline flink-sql-runner/src/test/java/com/github/streamshub/flink/KubernetesSecretReplacerImplTest.java:32]: I would create a variable/constant with `my-secret`, same for namespace

File: flink-sql-runner/src/test/java/com/github/streamshub/flink/SqlRunnerTest.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+import java.util.List;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class SqlRunnerTest {
+
+    @Test
+    void testParseStatements() {
+        String statements = "CREATE TABLE Table1 ( field STRING, field1 STRING ) WITH ( 'connector' = 'my-source-connector', 'path' = '/test1/data' ); CREATE TABLE Table2 ( field STRING, field2 STRING ) WITH ( 'connector' = 'my-sink-connector', 'path' = '/path/data2' ); INSERT INTO Table2 SELECT * FROM Table1;";
```
Comment (Frawless) [inline flink-sql-runner/src/test/java/com/github/streamshub/flink/SqlRunnerTest.java:12]: Just out of curiosity - isn't there some java library that offers builder for SQL statements?

Comment (robobario) [reply flink-sql-runner/src/test/java/com/github/streamshub/flink/SqlRunnerTest.java:12]: I had a play with jOOQ (which offers a [DSL for building generic SQL strings](https://www.jooq.org/doc/3.19/manual/sql-building/)) and I don't think it has any concept of the WITH properties style Flink uses to define the additional props for a table. Also you have to tinker to make sure it formats the output the way we like, and it's resistant to emitting "string" as the data type in the different SQL dialects I tried.

I think in this case we're better off with String manipulation since the runner cares a lot about splitting on delimiters and less about the SQL validity, we want to know exactly what strings are running through it. Though I think it's good to test with some realistic looking inputs.

File: flink-sql-runner/src/test/java/com/github/streamshub/flink/SqlRunnerTest.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
+package com.github.streamshub.flink;
+
+import java.util.List;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class SqlRunnerTest {
```
Comment (robobario) [inline flink-sql-runner/src/test/java/com/github/streamshub/flink/SqlRunnerTest.java:8]: Another behaviour to test is that `parseStatements` adds a trailing `;` if none is present.

Maybe something for another PR, but I'd refactor out the work in the SqlRunner main method to make it testable. We have this untested logic where set statements are put in the configuration of the table env rather than executed. You could separate the environment/execution logic from the statement processing to make it more testable.

File: flink-sql-runner/src/test/java/com/github/streamshub/flink/SqlRunnerTest.java (hunk 3 above)
Comment (tinaselenge) [reply flink-sql-runner/src/test/java/com/github/streamshub/flink/SqlRunnerTest.java:12]: Thanks @robobario for checking how jOOQ  works. In that case, I agree on using String manipulation rather a library for building SQL statements. I can update the SQL statements to look more realistic.

File: flink-sql-runner/src/test/java/com/github/streamshub/flink/SqlRunnerTest.java (hunk 4 above)
Comment (tinaselenge) [reply flink-sql-runner/src/test/java/com/github/streamshub/flink/SqlRunnerTest.java:8]: That's a good point. Given that we already diverged off a little bit from the upstream, I think we should refactor it to make it more testable.

File: flink-sql-runner/src/test/java/com/github/streamshub/flink/SqlRunnerTest.java (hunk 3 above)
Comment (Frawless) [reply flink-sql-runner/src/test/java/com/github/streamshub/flink/SqlRunnerTest.java:12]: That makes sense to me, thanks for explanation

File: flink-sql-runner/src/test/java/com/github/streamshub/flink/SqlRunnerTest.java (hunk 4 above)
Comment (tinaselenge) [reply flink-sql-runner/src/test/java/com/github/streamshub/flink/SqlRunnerTest.java:8]: I have added a test case for adding trailing semicolon to the end of the statements. I have left the refactor for another PR.

**Review by kornys — APPROVED** [review]

File: flink-sql-runner/src/test/java/com/github/streamshub/flink/SqlRunnerTest.java (hunk 4 above)
Comment (SamBarker) [reply flink-sql-runner/src/test/java/com/github/streamshub/flink/SqlRunnerTest.java:8]: Is it worth seeing if upstream would be interested in the testing changes?

**Review by SamBarker — APPROVED** [review]
Starting to add some coverage is good. As I prefer asserJ's assertions to the base ones in Junit I've opened https://github.com/tinaselenge/flink-sql/pull/2 to see what you think as well.

**Review by robobario — APPROVED** [review]
LGTM

File: flink-sql-runner/src/test/java/com/github/streamshub/flink/SqlRunnerTest.java (hunk 4 above)
Comment (tinaselenge) [reply flink-sql-runner/src/test/java/com/github/streamshub/flink/SqlRunnerTest.java:8]: Yes, potentially. I think because it's an example code, they didn't have any tests?

File: flink-sql-runner/src/main/java/com/github/streamshub/flink/KubernetesSecretReplacer.java [hunk 5]
```diff
@@ -1,46 +1,14 @@
...
-        } catch (KubernetesClientException e) {
-            throw new RuntimeException(e);
-        }
-    }
-}
\ No newline at end of file
+/**
+* An abstraction over a class that resolves Kubernetes secrets
+ */
+public interface KubernetesSecretReplacer {
```
[the remark this answers is not in the document]
Comment (tinaselenge) [reply flink-sql-runner/src/main/java/com/github/streamshub/flink/KubernetesSecretReplacer.java]: good idea!

File: flink-sql-runner/src/main/java/com/github/streamshub/flink/InterpolatorImpl.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+import io.fabric8.kubernetes.api.model.Secret;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.KubernetesClientBuilder;
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class InterpolatorImpl implements Interpolator {
```
Comment (SamBarker) [inline flink-sql-runner/src/main/java/com/github/streamshub/flink/InterpolatorImpl.java:11]: I would have kept the original name really. `KubernetesSecretInterpolator` ?

**Review by SamBarker — APPROVED** [review]

File: flink-sql-runner/src/main/java/com/github/streamshub/flink/InterpolatorImpl.java (hunk 6 above)
Comment (tinaselenge) [reply flink-sql-runner/src/main/java/com/github/streamshub/flink/InterpolatorImpl.java]: Yeah, that sounds better actually.

**Comment by tinaselenge** [comment]
> Starting to add some coverage is good. As I prefer asserJ's assertions to the base ones in Junit I've opened [tinaselenge#2](https://github.com/tinaselenge/flink-sql/pull/2) to see what you think as well.

Thanks @SamBarker .  This looks good to me. Do you want me to merge that into my branch before merging this PR or do you want to open a follow up PR against the main?

File: flink-sql-runner/src/main/java/com/github/streamshub/flink/KubernetesSecretInterpolator.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+    // Expected pattern for a secret is {{secret:<namespace>/<name>/<key>}}
+    private static final Pattern SECRET_PATTERN = Pattern.compile("\\{\\{secret:([^/]+)/([^}]+)/([^}]+)}}");
+
+    public KubernetesSecretInterpolator() {
+        this.client = new KubernetesClientBuilder().build();
+    }
+
+    // visible for testing
+    public KubernetesSecretInterpolator(KubernetesClient client) {
```
Comment (katheris) [inline flink-sql-runner/src/main/java/com/github/streamshub/flink/KubernetesSecretInterpolator.java:22]: Could this be protected rather than public? I also quite like the Strimzi convention of prepending methods like this with `/* test */` rather than having the comment on a separate line

**Review by katheris — APPROVED** [review]
Generally looks good to me, just one small nit.

Also more generally I wonder if it's worth thinking about test naming. I saw something recently about using given_when_then() names to make test names more descriptive. For example instead of `testInterpolateSecrets()` have `givenStatementSecrets_whenInterpolate_thenSecretValuesInserted()`. It's a much longer test name but once you get beyond a small number of tests it gets harder and harder to name them. What do people think?

**Comment by katheris** [comment]
> Starting to add some coverage is good. As I prefer asserJ's assertions to the base ones in Junit I've opened [tinaselenge#2](https://github.com/tinaselenge/flink-sql/pull/2) to see what you think as well.

+1 from me for using assertJ

**Comment by Frawless** [comment]
> Generally looks good to me, just one small nit.
> 
> Also more generally I wonder if it's worth thinking about test naming. I saw something recently about using given_when_then() names to make test names more descriptive. For example instead of `testInterpolateSecrets()` have `givenStatementSecrets_whenInterpolate_thenSecretValuesInserted()`. It's a much longer test name but once you get beyond a small number of tests it gets harder and harder to name them. What do people think?

wouldn't it cause problems with line length at some point? I mean with complicated tests it might be tricky (probably it more applies to some integration/system tests)

> +1 from me for using assertJ
Just out of curiosity - are there some significant benefits from assertJ that JUnit doesn't provide? We used hamcrest matchers in Strimzi, but in new projects are kinda stick with JUnit to avoid another test dependency and jUnit works fine for our use-cases.

**Comment by SamBarker** [comment]
> Just out of curiosity - are there some significant benefits from assertJ that JUnit doesn't provide? We used hamcrest matchers in Strimzi, but in new projects are kinda stick with JUnit to avoid another test dependency and jUnit works fine for our use-cases.

Yes. For me it's superpower is non technical it allows me to read the assertions more like a sentence so that when reading test code it's much easier to understand what is trying to be demonstrated. 

For the simpler assertions there isn't much difference between `assertTrue(booleanThing)` and `assertThat(booleanThing).isTrue()`. When we get to more involved assertions involving multiple things such as method x throws an exception of type Y with message Z. Take a look at https://github.com/kroxylicious/kroxylicious/blob/main/kroxylicious-filters/kroxylicious-record-encryption/src/test/java/io/kroxylicious/filter/encryption/dek/DekManagerTest.java#L111-L115 (just because I was reviewing it yesterday) which asserts that a future has failed with a time period and then in the same assertion validates properties of the failure.

That’s all possible to do with JUnit and hamcrest but with a lot more code and usually reimplemented for each test


@tinaselenge i don't mind happy to covert my change to one pointed at main once this is merged if that moves quicker

**Comment by tinaselenge** [comment]
Thanks everyone for reviewing the PR! 

> > Generally looks good to me, just one small nit.
> > Also more generally I wonder if it's worth thinking about test naming. I saw something recently about using given_when_then() names to make test names more descriptive. For example instead of `testInterpolateSecrets()` have `givenStatementSecrets_whenInterpolate_thenSecretValuesInserted()`. It's a much longer test name but once you get beyond a small number of tests it gets harder and harder to name them. What do people think?
> 
> wouldn't it cause problems with line length at some point? I mean with complicated tests it might be tricky (probably it more applies to some integration/system tests)

Given that we don't have that many unit tests and methods to test for now, I think the current naming is ok. I agree that the descriptive naming perhaps works well for integration/system.  I will leave this discussion for a future PR to merge this PR now.
